internal/objects/contour/contour.go

```diff
@@ -0,0 +1,123 @@
+// namespace/name, pubType for the network publishing type of Envoy, and
+// Envoy container ports 8080/8443.
+func New(name, ns, specNs string, remove bool, pubType operatorv1alpha1.NetworkPublishingType) *operatorv1alpha1.Contour {
+	return &operatorv1alpha1.Contour{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      name,
+		},
+		Spec: operatorv1alpha1.ContourSpec{
+			Namespace: operatorv1alpha1.NamespaceSpec{
+				Name:             specNs,
+				RemoveOnDeletion: remove,
+			},
+			NetworkPublishing: operatorv1alpha1.NetworkPublishing{
+				Envoy: operatorv1alpha1.EnvoyNetworkPublishing{
+					Type: pubType,
+					ContainerPorts: []operatorv1alpha1.ContainerPort{
+						{
+							Name:       "http",
+							PortNumber: int32(8080),
+						},
+						{
+							Name:       "https",
+							PortNumber: int32(8443),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// OwnerLabelsExist returns true if obj contains Contour owner labels.
+func OwnerLabelsExist(obj metav1.Object, contour *operatorv1alpha1.Contour) bool {
+	labels := obj.GetLabels()
+	nameFound := false
+	nsFound := false
+	if labels == nil {
+		return false
+	}
+	for l, v := range labels {
+		switch {
+		case nameFound && nsFound:
+			return true
+		case l == operatorv1alpha1.OwningContourNameLabel && v == contour.Name:
+			nameFound = true
+		case l == operatorv1alpha1.OwningContourNsLabel && v == contour.Namespace:
+			nsFound = true
+		}
+	}
+	if nameFound && nsFound {
+		return true
+	}
+	// no contour owning name and ns labels found.
+	return false
+}
+
+// OtherContoursExist lists Contour objects in all namespaces, returning the list
+// and true if any exist other than contour.
+func OtherContoursExist(ctx context.Context, cli client.Client, contour *operatorv1alpha1.Contour) (bool, *operatorv1alpha1.ContourList, error) {
+	contours := &operatorv1alpha1.ContourList{}
+	if err := cli.List(ctx, contours); err != nil {
+		return false, nil, fmt.Errorf("failed to list contours: %w", err)
+	}
+	if len(contours.Items) == 0 || len(contours.Items) == 1 && contours.Items[0].Name == contour.Name {
+		return false, nil, nil
+	}
+	return true, contours, nil
+}
+
+// OtherContoursExistInSpecNs lists Contour objects in the same spec.namespace.name as contour,
+// returning true if any exist.
+func OtherContoursExistInSpecNs(ctx context.Context, cli client.Client, contour *operatorv1alpha1.Contour) (bool, error) {
+	exist, contours, err := OtherContoursExist(ctx, cli, contour)
+	if err != nil {
+		return false, err
+	}
+	if exist {
+		for _, c := range contours.Items {
+			if c.Spec.Namespace.Name == contour.Spec.Namespace.Name {
+				return true, nil
+			}
+		}
+	}
+	return false, nil
+}
+
+// OwningSelector returns a label selector using "contour.operator.projectcontour.io/owning-contour-name"
+// and "contour.operator.projectcontour.io/owning-contour-namespace" labels.
+func OwningSelector(contour *operatorv1alpha1.Contour) *metav1.LabelSelector {
```

Comment:
This could refactor into an instance method as well.

Reply:
Same response as https://github.com/projectcontour/contour-operator/pull/204#discussion_r578085255.